### PR TITLE
Poll Plugin: Add Voting History

### PR DESCRIPTION
**Pull request by xichi**
CAU3@@G`$OT~XOZ7FZ1FJ)4
175W~X9MCE6D4U8B2YZG6BG

### CODE REVIEW

File: src/plugins/Polls/Services.ts [hunk 1]
```diff
@@ -20,16 +20,16 @@ const defaultPoll: PollGunDB = {
 }
 
 interface NewPollProps {
-    sender?: string | undefined
-    id?: string | undefined
+    sender_name?: string | undefined
+    sender_id?: string | undefined
```
Comment (septs) [inline src/plugins/Polls/Services.ts:24]:
```suggestion
    sender_name?: string
    sender_id?: string
```

File: src/plugins/Polls/schema.json [hunk 2]
```diff
@@ -26,14 +26,23 @@
             "type": "array",
             "title": "results"
         },
-        "sender": {
+        "sender_name": {
```
Comment (septs) [inline src/plugins/Polls/schema.json:29]: change schema is break change.

v1.16.2 include the plugin on release.

CC @guanbinrui @Jack-Works

Comment (Jack-Works) [reply src/plugins/Polls/schema.json:29]: Agree. You need to make it back-ward compatibility or add a new version.

File: src/plugins/Polls/types.ts [hunk 3]
```diff
@@ -5,8 +5,16 @@ export interface PollMetaData {
     end_time: number
     options: string[]
     results: number[]
-    sender?: string
-    id?: string
+    sender_name?: string
+    sender_id?: string
+    voting_history?: VotingDetail[]
```
Comment (Jack-Works) [inline src/plugins/Polls/types.ts:10]: This seems strange. How could you upgrade the data that has been serialized into the post?

File: src/plugins/Polls/schema.json (hunk 2 above)
Comment (septs) [reply src/plugins/Polls/schema.json:29]: i think, we can next release disable this plugin in production.

it is data structure not stable plugin.

**Comment by 1999-XiChi** [comment]
@septs @Jack-Works 
What do I need to do for the new features now？

**Comment by septs** [comment]
CC @yisiliu

**Comment by Jack-Works** [comment]
this plugin relies on the gun but the gun is offline now (IIRC) what to do? @septs

**Comment by Tedko** [comment]
> this plugin relies on the gun but the gun is offline now (IIRC) what to do? @septs

seems gun's back?

**Comment by Tedko** [comment]
seems a bug:

**Comment by Tedko** [comment]
> seems a bug:

So basically one user tell me that hes able to click 100times of one option and after refresh the option (100times) disappreaed.

Should be a known one

**Review by 1999-XiChi — COMMENTED** [review]
> > seems a bug:
> 
> So basically one user tell me that hes able to click 100times of one option and after refresh the option (100times) disappreaed.
> 
> Should be a known one

I will fix it.

File: packages/maskbook/src/plugins/Polls/UI/Polls.tsx [hunk 4]
```diff
@@ -55,23 +58,32 @@ const useStyles = makeStyles((theme) =>
...
             color: '#657786',
         },
+        history: {
+            color: theme.palette.primary.main,
+        },
     }),
 )
 
-interface PollCardProps {
+interface PollCardProps extends withClasses<KeysInferFromUseStyles<typeof useStyles> | 'button'> {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Polls/UI/Polls.tsx:67]: Hi `KeysInferFromUseStyles` has been deprecated please avoid of using this

File: packages/maskbook/src/plugins/Polls/UI/Polls.tsx [hunk 5]
```diff
@@ -55,23 +58,32 @@ const useStyles = makeStyles((theme) =>
...
+    const { poll, onClick, onOpenHistoryDialog, vote, status } = props
+    const classes = useStylesExtends(useStyles(), props)
+    const [historyOpen, setHistoryOpen] = useState(false)
     const isClosed = Date.now() > poll.end_time ? true : false
     const { t } = useI18N()
     const lang = useValueRef(languageSettings)
 
+    useEffect(() => {
+        onOpenHistoryDialog?.(historyOpen)
+    }, [historyOpen])
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Polls/UI/Polls.tsx:85]: Also add `onOpenHistoryDialog` to deps

File: packages/maskbook/src/plugins/Polls/schema.json [hunk 6]
```diff
@@ -26,14 +26,23 @@
             "type": "array",
             "title": "results"
         },
-        "sender": {
+        "sender_name": {
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Polls/schema.json:29]: This is a breaking change

File: packages/maskbook/src/plugins/Polls/UI/VotingHistory.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        content: {
+            paddingBottom: '20px',
+            paddingTop: '0px',
+        },
+    }),
+)
+
+interface VotingHistoryProps
+    extends withClasses<
+        | KeysInferFromUseStyles<typeof useStyles>
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Polls/UI/VotingHistory.tsx:31]: `KeysInferFromUseStyles` is deprecated

File: packages/maskbook/src/plugins/Polls/UI/VotingHistory.tsx [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+export default function VotingHistory(props: VotingHistoryProps) {
+    const { poll } = props
+    const { voting_history, options } = poll
+    const classes = useStylesExtends(useStyles(), props)
+    const { t } = useI18N()
+
+    return (
+        <>
+            <ShadowRootDialog
```
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Polls/UI/VotingHistory.tsx:53]: Please migrate to new cleaner API `InjectedDialog`

**Comment by Tedko** [comment]
hi @1999-XiChi @Jack-Works  any update?

**Comment by 1999-XiChi** [comment]
> hi @1999-XiChi @Jack-Works any update?

Recently busy taking driving lessons, I'll submit update this weekend

**Comment by Tedko** [comment]
> > hi @1999-XiChi @Jack-Works any update?
> 
> Recently busy taking driving lessons, I'll submit update this weekend

cool!

**Comment by Tedko** [comment]
@1999-XiChi hihi ~ any updates:)? 

btw wanna send you some swags as well

**Pull request closed without merging**